File: src/Arrays/subArrayWithXorK/SubArrayWithXorK.js
module.exports = { 
 //param A : array of integers
 //param B : integer
 //return an integer
	solve : function(A, B){
        // TC: O(N^3)
        // let n = A.length, count = 0;
        // for (let i = 0; i < n; i++) {
        //     for (let j = i; j < n; j++) {
        //         let xor = 0;
        //         for (let k = i; k <= j; k++) {
        //             xor = xor ^ A[k];
        //         }
        //         if (xor === B)
        //             count++;
        //     }
        // }
        // return count;
        
        // TC: O(N^2)
        // let n = A.length, count = 0;
        // for (let i = 0; i < n; i++) {
        //     let xor = 0;
        //     for (let j = i; j < n; j++) {
        //         xor = xor ^ A[j];
        //         if (xor === B) 
        //             count++;
        //     }
        // }
        // return count;
        
        // TC: O(N) or O(N*logN)
        // SC: O(N)
        let map = new Map();
        map.set(0, 1);
        let count = 0, xor = 0;
        for (let a of A) {
            xor = xor ^ a;
            let x = xor ^ B;
            count = count + (map.get(x) || 0);
            map.set(xor, (map.get(xor) || 0) + 1);
        }
        return count;
	}
};

File: src/Arrays/subArrayWithXorK/SubarrayWithXorK.java
public class Solution {
    public int solve(ArrayList<Integer> A, int B) {
        // TC: O(N) or O(N*log N)
        // SC: O(N)
        Map<Integer, Integer> mpp = new HashMap<>();
        mpp.put(0, 1);
        int count = 0, xor = 0;
        for (int a : A) {
            xor = xor ^ a;
            int preXor = xor ^ B;
            count += mpp.getOrDefault(preXor, 0);
            mpp.put(xor, mpp.getOrDefault(xor, 0) + 1);
        }
        return count; 
    }
}
